From f71258eaf85eebce4e857f1bfcad18bc3f4c2442 Mon Sep 17 00:00:00 2001
From: Chen Zhongjin <chenzhongjin@huawei.com>
Date: Wed, 31 Aug 2022 16:25:03 +0800
Subject: [PATCH 881/905] [Huawei] bpf: Add helper bpf_update_tcp_seq to
 synchronize tcp seq/ack

Offering: HULK
hulk inclusion
category: feature
bugzilla: N/A

--------------------------------

In order to process TCP packet with BPF/XDP, it's necessary to
synchronize seq and ack between kernel network stack and bpf prog.
This patch introduces a sample helper to do the sync.

Note that, it's only used for samples, retransmission and congestion
control are not supported.

Add compile macro BPF_XDP_UPDATE_TCP_SEQ_HELPER to control whether
the helper bpf_update_tcp_seq is enabled.

Signed-off-by: Chen Zhongjin <chenzhongjin@huawei.com>
Signed-off-by: He Fengqing <hefengqing@huawei.com>
Signed-off-by: Xu Kuohai <xukuohai@huawei.com>
Signed-off-by: Yang Jihong <yangjihong@huawei.com>
---
 include/uapi/linux/bpf.h       | 13 +++++++
 net/Kconfig                    |  9 +++++
 net/core/filter.c              | 70 ++++++++++++++++++++++++++++++++++
 tools/include/uapi/linux/bpf.h | 13 +++++++
 4 files changed, 105 insertions(+)

diff --git a/include/uapi/linux/bpf.h b/include/uapi/linux/bpf.h
index bd566bfc843f..8f3dd974517b 100644
--- a/include/uapi/linux/bpf.h
+++ b/include/uapi/linux/bpf.h
@@ -3755,6 +3755,14 @@ union bpf_attr {
  *             Get Ipv4 origdst or replysrc. Works with IPv4.
  *     Return
  *             0 on success, or a negative error in case of failure.
+ *
+ * int bpf_update_tcp_seq(struct xdp_buff *ctx, struct bpf_sock_tuple *tuple, u32 len, u32 netns_id, u64 flags)
+ *     Description
+ *             Update tcp seq, only available with kernel compiled with
+ *             the **CONFIG_BPF_XDP_UPDATE_TCP_SEQ_HELPER** configuration
+ *             option set to "**y**".
+ *     Return
+ *             0 on success, or a negative error in case of failure.
  */
 #define __BPF_FUNC_MAPPER(FN)		\
 	FN(unspec),			\
@@ -3915,6 +3923,7 @@ union bpf_attr {
 	FN(redirect_peer),		\
 	FN(get_sockops_uid_gid),	\
 	FN(sk_original_addr),		\
+	FN(update_tcp_seq),		\
 	/* */
 
 /* integer value in 'imm' field of BPF_CALL instruction selects which helper
@@ -4255,6 +4264,10 @@ struct bpf_sock_tuple {
 			__be32 daddr;
 			__be16 sport;
 			__be16 dport;
+			__be32 seq;
+			__be32 rcv_len;
+			__be32 ack_seq;
+			__be32 snd_len;
 		} ipv4;
 		struct {
 			__be32 saddr[4];
diff --git a/net/Kconfig b/net/Kconfig
index 0bd8fdc6f58d..db3cb5e52cd8 100644
--- a/net/Kconfig
+++ b/net/Kconfig
@@ -330,6 +330,15 @@ config EULER_SOCKETMAP
 	help
 	  Enabling this support socket map in EulerOS.
 
+config BPF_XDP_UPDATE_TCP_SEQ_HELPER
+	bool "enable XDP helper bpf_update_tcp_seq"
+	default n
+	help
+	  Enable ebpf helper to allows XDP synchronize the TCP socket status.
+	  This helper will update socket status directly only with minimal
+	  essential check so it is caller's responsibility to promise the
+	  correctness of parameters.
+
 config NET_FLOW_LIMIT
 	bool
 	depends on RPS
diff --git a/net/core/filter.c b/net/core/filter.c
index 2a5e13a298b5..8eec132ed452 100644
--- a/net/core/filter.c
+++ b/net/core/filter.c
@@ -6276,6 +6276,72 @@ static const struct bpf_func_proto bpf_xdp_sk_lookup_tcp_proto = {
 	.arg5_type      = ARG_ANYTHING,
 };
 
+#ifdef CONFIG_BPF_XDP_UPDATE_TCP_SEQ_HELPER
+/* If we update tp->rcv_nxt, also update tp->bytes_received */
+static inline void tcp_rcv_nxt_update(struct tcp_sock *tp, u32 seq, u32 rcv_len)
+{
+	u32 next_seq = seq + rcv_len;
+
+	tp->bytes_received += rcv_len;
+	WRITE_ONCE(tp->rcv_nxt, next_seq);
+	WRITE_ONCE(tp->copied_seq, next_seq);
+}
+
+static inline void tcp_snd_nxt_update(struct tcp_sock *tp, u32 ack_seq, u32 snd_len)
+{
+	u32 next_ack = ack_seq + snd_len;
+
+	tp->bytes_sent += snd_len;
+	WRITE_ONCE(tp->snd_nxt, next_ack);
+	WRITE_ONCE(tp->write_seq, next_ack);
+}
+
+BPF_CALL_5(bpf_xdp_update_tcp_seq, struct xdp_buff *, ctx,
+	   struct bpf_sock_tuple *, tuple, u32, len, u32, netns_id, u64, flags)
+{
+	struct net *caller_net = dev_net(ctx->rxq->dev);
+	int ifindex = ctx->rxq->dev->ifindex;
+	struct sock *sk;
+	struct tcp_sock *tp;
+	u32 seq, ack;
+
+	sk = __bpf_sk_lookup(NULL, tuple, len, caller_net,
+			     ifindex, IPPROTO_TCP, netns_id,
+			     flags);
+	if (!sk)
+		return -1;
+
+	tp = tcp_sk(sk);
+
+	seq = tuple->ipv4.seq;
+	ack = tuple->ipv4.ack_seq;
+
+	if (seq == tp->copied_seq && seq == tp->rcv_nxt &&
+	    ack == tp->write_seq && ack == tp->snd_nxt) {
+		sock_owned_by_me((struct sock *)tp);
+		tp->bytes_acked = tp->bytes_sent;
+		tcp_rcv_nxt_update(tp, seq, tuple->ipv4.rcv_len);
+		tcp_snd_nxt_update(tp, ack, tuple->ipv4.snd_len);
+
+		return 0;
+	}
+
+	return -1;
+}
+
+static const struct bpf_func_proto bpf_xdp_update_tcp_seq_proto = {
+	.func           = bpf_xdp_update_tcp_seq,
+	.gpl_only       = false,
+	.pkt_access     = true,
+	.ret_type       = RET_INTEGER,
+	.arg1_type      = ARG_PTR_TO_CTX,
+	.arg2_type      = ARG_PTR_TO_MEM,
+	.arg3_type      = ARG_CONST_SIZE,
+	.arg4_type      = ARG_ANYTHING,
+	.arg5_type      = ARG_ANYTHING,
+};
+#endif
+
 BPF_CALL_5(bpf_sock_addr_skc_lookup_tcp, struct bpf_sock_addr_kern *, ctx,
 	   struct bpf_sock_tuple *, tuple, u32, len, u64, netns_id, u64, flags)
 {
@@ -7354,6 +7420,10 @@ xdp_func_proto(enum bpf_func_id func_id, const struct bpf_prog *prog)
 		return &bpf_xdp_sk_lookup_udp_proto;
 	case BPF_FUNC_sk_lookup_tcp:
 		return &bpf_xdp_sk_lookup_tcp_proto;
+#ifdef CONFIG_BPF_XDP_UPDATE_TCP_SEQ_HELPER
+	case BPF_FUNC_update_tcp_seq:
+		return &bpf_xdp_update_tcp_seq_proto;
+#endif
 	case BPF_FUNC_sk_release:
 		return &bpf_sk_release_proto;
 	case BPF_FUNC_skc_lookup_tcp:
diff --git a/tools/include/uapi/linux/bpf.h b/tools/include/uapi/linux/bpf.h
index 6edb480050b5..639d6aafd85e 100644
--- a/tools/include/uapi/linux/bpf.h
+++ b/tools/include/uapi/linux/bpf.h
@@ -3755,6 +3755,14 @@ union bpf_attr {
  *             Get Ipv4 origdst or replysrc. Works with IPv4.
  *     Return
  *             0 on success, or a negative error in case of failure.
+ *
+ * int bpf_update_tcp_seq(struct xdp_buff *ctx, struct bpf_sock_tuple *tuple, u32 len, u32 netns_id, u64 flags)
+ *     Description
+ *             Update tcp seq, only available with kernel compiled with
+ *             the **CONFIG_BPF_XDP_UPDATE_TCP_SEQ_HELPER** configuration
+ *             option set to "**y**".
+ *     Return
+ *             0 on success, or a negative error in case of failure.
  */
 #define __BPF_FUNC_MAPPER(FN)		\
 	FN(unspec),			\
@@ -3915,6 +3923,7 @@ union bpf_attr {
 	FN(redirect_peer),		\
 	FN(get_sockops_uid_gid),	\
 	FN(sk_original_addr),		\
+	FN(update_tcp_seq),		\
 	/* */
 
 /* integer value in 'imm' field of BPF_CALL instruction selects which helper
@@ -4254,6 +4263,10 @@ struct bpf_sock_tuple {
 			__be32 daddr;
 			__be16 sport;
 			__be16 dport;
+			__be32 seq;
+			__be32 rcv_len;
+			__be32 ack_seq;
+			__be32 snd_len;
 		} ipv4;
 		struct {
 			__be32 saddr[4];
-- 
2.17.1

